aips/aip-20.md

```diff
@@ -0,0 +1,399 @@
+- `neg()` for inverse.
+- `add()` for a basic group operation.
+- `sub()` for group element subtraction.
+- `double()` for efficient doubling.
+- `scalar_mul()` for group scalar multiplication.
+- `multi_scalar_mul()` for efficient group multi-scalar multiplication.
+- `hash_to()` for hash-to-group.
+
+A field may also implement the following operations.
+- `zero()` for the field additive identity.
+- `one()` for the field multiplicative identity.
+- `add()` for field addition.
+- `sub()` for field subtraction.
+- `mul()` for field multiplication.
+- `div()` for field division.
+- `neg()` for field negation.
+- `inv()` for field inversion.
+- `sqr()` for efficient field element squaring.
+- `from_u64()` for quick conversion from u64 to field element.
+
+For 3 groups that form a bilinear map, `pairing()` and `multi_pairing()` may be implemented.
+
+For a subset/superset relationship between 2 structures, `upcast()` and `downcast()` may be implemented.
+
+#### Shared Scalar Fields
+
+Some groups share the same group order,
+and an ergonomic design from this is to
+allow multiple groups to share the same scalar field
+(mainly for the purpose scalar multiplication),
+if they have the same order.
+
+E.g. the following should be supported.
+
+```rust
+// algebra.move
+pub fun scalar_mul<G,S>(element: &Element<G>, scalar: &Scalar<S>)...
+
+// user_contract.move
+**let k: Scalar<ScalarForBx> = somehow_get_k();**
+let p1 = one<GroupB1>();
+let p2 = one<GroupB2>();
+let q1 = scalar_mul<GroupB1, ScalarForBx>(&p1, &k);
+let q2 = scalar_mul<GroupB2, ScalarForBx>(&p2, &k);
+```
+
+#### Handling Incorrect Type Parameter(s)
+
+There is currently no easy way to ensure type safety for the generic operations.
+E.g., `pairing<A,B,C>(a,b,c)` can compile even there is no pairing between `A,B,C` existing/implemented.
+
+Therefore the backend should handle the type checks at runtime.
+
+If a group operations that takes 2+ type parameters is invoked with incompatible type parameters, it should abort.
+E.g. `scalar_mul<GroupA, ScalarForBx>()` should abort with a “not implemented” error.
+
+Invoking operation functions with user-defined types should also abort with a “not implemented” error.
+
+### Implementation of BLS12-381 structures
+
+The construction of BLS12-381 curves involve many groups/fields, some frequently interacted by applications (e.g., `Fq12`, `Fr`, `G1Affine`, `G2Affine`, `Gt`) while others rarely used. Marker types for using these structures with `aptos_std::algebra` APIs should be defined and exposed to developers, along with their widely-used serialization formats and hash-to-group suites, (ideally in its own module named `aptos_std::algebra_bls12381`).
+
+Below are the full specification in pseudo-Move.
+
+NOTE: some items below are marked "not implemented" but still presented here to facilitate the definition of some other items.
+
+```rust
+module aptos_std::algebra_bls12381 {
+    /// The finite field $F_q$ used in BLS12-381 curves.
```

Comment:
`F_q`

Reply:
It's intended latex. I also rewrote the bls12381 section so those equations can render.